CMakeLists.txt

```diff
@@ -0,0 +1,290 @@
+if(EMSCRIPTEN)
+    message(STATUS "Trace logging disabled in wasm builds.")
+    set(ENABLE_TRACE OFF)
+elseif(EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/src/rom_lst.h" AND
+       EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/src/rom_labels.h")
+    set(ENABLE_TRACE ON)
+    message(STATUS "Trace headers found (rom_lst.h, rom_labels.h). Enabling TRACE.")
+else()
+    set(ENABLE_TRACE OFF)
+    message(STATUS "Trace headers not found. TRACE disabled.")
+endif()
+
+# --- Code Generation (Python Script) ---
+# This replaces the 'cpu/tables.h' rule in the Makefile
+set(TABLE_GEN_SCRIPT "${CMAKE_CURRENT_SOURCE_DIR}/src/cpu/buildtables.py")
+set(TABLE_OUTPUT_H "${CMAKE_CURRENT_SOURCE_DIR}/src/cpu/tables.h")
+set(MNEMONIC_OUTPUT_H "${CMAKE_CURRENT_SOURCE_DIR}/src/cpu/mnemonics.h")
+
+add_custom_command(
+    OUTPUT ${TABLE_OUTPUT_H} ${MNEMONIC_OUTPUT_H}
+    COMMAND "${Python3_EXECUTABLE}" "${TABLE_GEN_SCRIPT}"
+    WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/src/cpu"
+    DEPENDS "${TABLE_GEN_SCRIPT}" "${CMAKE_CURRENT_SOURCE_DIR}/src/cpu/6502.opcodes"
+    DEPENDS "${TABLE_GEN_SCRIPT}" "${CMAKE_CURRENT_SOURCE_DIR}/src/cpu/65c02.opcodes"
+    DEPENDS "${TABLE_GEN_SCRIPT}" "${CMAKE_CURRENT_SOURCE_DIR}/src/cpu/65c816.opcodes"
+    COMMENT "Generating CPU tables..."
+)
+# Create a target so the main executable can depend on it
+add_custom_target(gen_cpu_tables DEPENDS ${TABLE_OUTPUT_H})
+
+# --- Sources ---
+set(X16_SOURCES
+    src/cpu/fake6502.c
+    src/memory.c
+    src/disasm.c
+    src/video.c
+    src/i2c.c
+    src/smc.c
+    src/rtc.c
+    src/via.c
+    src/serial.c
+    src/ieee.c
+    src/vera_spi.c
+    src/audio.c
+    src/vera_pcm.c
+    src/vera_psg.c
+    src/sdcard.c
+    src/main.c
+    src/debugger.c
+    src/javascript_interface.c
+    src/joystick.c
+    src/rendertext.c
+    src/keyboard.c
+    src/icon.c
+    src/timing.c
+    src/wav_recorder.c
+    src/testbench.c
+    src/files.c
+    src/cartridge.c
+    src/iso_8859_15.c
+    src/ymglue.cpp
+    src/midi.c
+    src/extern/ymfm/src/ymfm_opm.cpp
+)
+
+if(WIN32)
+    list(APPEND X16_SOURCES src/video_win32.c)
+endif()
+
+# --- Main Executable Target ---
+add_executable(x16emu ${X16_SOURCES})
+
+# Ensure the Python script runs before compiling x16emu
+add_dependencies(x16emu gen_cpu_tables)
+
+# Includes
+target_include_directories(x16emu PRIVATE
+    src
+    src/extern/include
+    src/extern/ymfm/src
+    ${SDL2_INCLUDE_DIRS}
+)
+
+if(ENABLE_TRACE)
+    target_compile_definitions(x16emu PRIVATE TRACE)
+endif()
+
+# Linking
+target_link_libraries(x16emu PRIVATE
+    ${SDL2_LIBRARIES}
+    ZLIB::ZLIB
+)
+
+if(FLUIDSYNTH_FOUND)
+    target_link_libraries(x16emu PRIVATE PkgConfig::FLUIDSYNTH)
+endif()
+
+# --- Platform Specific Configuration ---
+
+# 1. Link Time Optimization
+if(ENABLE_LTO)
+    include(CheckIPOSupported)
+    check_ipo_supported(RESULT result)
+    if(result)
+        set_property(TARGET x16emu PROPERTY INTERPROCEDURAL_OPTIMIZATION TRUE)
```

Comment:
Should rather be `CMAKE_INTERPRODECURAL_OPTIMIZATION` set before all the targets, imo.